modules/ml/src/main/java/org/apache/ignite/ml/naivebayes/bernoulli/BernoulliNaiveBayesTrainer.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.ml.naivebayes.bernoulli;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.ignite.ml.dataset.Dataset;
+import org.apache.ignite.ml.dataset.DatasetBuilder;
+import org.apache.ignite.ml.dataset.UpstreamEntry;
+import org.apache.ignite.ml.dataset.primitive.context.EmptyContext;
+import org.apache.ignite.ml.math.functions.IgniteBiFunction;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+import org.apache.ignite.ml.trainers.SingleLabelDatasetTrainer;
+
+/**
+ * Trainer for the Bernoulli naive Bayes classification model. The trainer calculates prior probabilities from the input
+ * binary dataset. Prior probabilities can be also set by {@code setPriorProbabilities} or {@code
+ * withEquiprobableClasses}. If {@code equiprobableClasses} is set, the probalilities of all classes will be {@code
+ * 1/k}, where {@code k} is classes count.
+ */
+public class BernoulliNaiveBayesTrainer extends SingleLabelDatasetTrainer<BernoulliNaiveBayesModel> {
+
+    /* Preset prior probabilities. */
+    private double[] priorProbabilities;
+    /* Sets equivalent probability for all classes. */
+    private boolean equiprobableClasses;
+    /** The threshold to convert a feature to a binary value. Default value is {@code 0.5}. */
+    private double binarizeThreshold = .5;
```

Comment:
I think that binarizeThreshold for all features in dataset is too strong assumption. Maybe you can add some preprocessing step that compute such threshold for each feature. 

Moreover such threshold explicitly sets that all features may be splitted into two classes when in general case feature distributions are fitted more precisely by histograms with several buckets (classes).

Reply:
@dehasi Excuse me for my previous comment. I was wrong and I hurried. You mean Bernoulli distribution in terms of feature distributions. From such point of view you're absolutely right but you can generalize this code to multivalue distributions (using hists) for features without functionality lost. convenient.

Reply:
let me just resolve this discussion. I added some relevant comment in the discussion below.